ros_control/RosTrajectoryExecutor

gilwoolee opened the pull request:
Need to review and write unit-tests. 

I have some issue compiling this. (I can only compile it if I comment out test-relevant lines in root CMakeLists.txt.)

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,79 @@
+// actionlib and DART both #define this macro.
+#undef DEPRECATED
+#include <actionlib/client/action_client.h>
+#undef DEPRECATED
```
@jslee02 Is this still necessary? I believe DART switched to the `DART_DEPRECATED` macro to avoid a conflict just like this.

jslee02 replied:
Not necessary. DART switched to the `DART_DEPRECATED` since version 6.0.1.

mkoval replied:
@gilwoolee Can you switch from `DEPRECATED` to `DART_GENERATED` and remove this `#undef`.

mkoval commented on src/control/ros/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+add_component_targets(${PROJECT_NAME} control_ros "${PROJECT_NAME}_control_ros")
+add_component_dependencies(${PROJECT_NAME} control_ros control statespace trajectory)
+
+# coveralls_add_sources(${sources})
```
Unless there is a strong reason not to, we should uncomment this.

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+      return "goal tolerance violated";
+
+    default:
+      return "unknown";
```
Nit: Include `errorCode` here, e.g.:
```c++
std::stringstream errorString;
errorString << "unknown (" << errorCode << ")";
return errorString.str()
```

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,79 @@
+    std::chrono::milliseconds connectionTimeout
+      = std::chrono::milliseconds{1000},
+    std::chrono::milliseconds connectionPollingRate
+      = std::chrono::milliseconds{20}
```
Nit: Rename this to `connectionPollingPeriod`.

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+RosTrajectoryExecutor::~RosTrajectoryExecutor()
+{
+  // Do nothing.
+  // TODO: Should we wait for the current trajectory to finish executing?
```
@psigen @ClintLiddick Any thoughts here?

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+  goal.goal_time_tolerance = ::ros::Duration(mGoalTimeTolerance);
+
+  // Convert the Aikido trajectory into a ROS JointTrajectory.
+  // TODO: Move this into a helper function.
```
We should move this into a helper function. Otherwise, it will be extremely difficult to unit test this function.

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+  util::StepSequence timeSequence{mTimestep, true, 0., _traj->getDuration()};
+  const auto numDofs = mSkeleton->getNumDofs();
+  const auto numWaypoints = timeSequence.getMaxSteps();
+  const auto numDerivatives = std::min<int>(_traj->getNumDerivatives(), 2);
```
I am not sure if this is correct - it seems like it should be valid for `numDerivatives` to be `1`, so we can execute a piecewise linear trajectory.

mkoval replied:
Good point. 👍

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+    Eigen::VectorXd tangentVector;
+    auto state = space->createState();
+    _traj->evaluate(timeAbsolute, state);
+    space->logMap(state, tangentVector);
```
Using `logMap` here only makes sense on `Rn` and `SO2` joints. We may want to add an explicit check to that effect. This is the compliment to the single-DOF restriction we have in #147.

gilwoolee replied:
Are we supporting anything other than `Rn` and `SO2` joints? I can add a type-check before the whole conversion happens.

mkoval replied:
I don't think so. At least, I can't think of any other joints that admit a *unique* parameterization as a single real number. Anything else is difficult to shoehorn into the `JointTrajectory` message.

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+  ROS_INFO("Transition: CommState = %s TerminalState = %s",
+    _handle.getCommState().toString().c_str(),
+    _handle.getTerminalState().toString().c_str());
+#endif
```
Delete dead code.

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+    {
+      mPromise.set_exception(
+        std::make_exception_ptr(
+          std::runtime_error(message.str())));
```
We should use a custom `std::runtime_error` subclass here. Otherwise, it will be difficult to distinguish between "expected" (e.g. trajectory was aborted) and "unexpected" failures (e.g. an invalid parameter).

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+    mInProgress = false;
+  }
+
+  // TODO: Check mGoalTimeTolerance here as a fail-safe.
```
I don't know what this means and don't think it is necessary.

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,79 @@
+    trajectory::TrajectoryPtr _traj, const ::ros::Time& _startTime);
+
+  /// Simulates mTraj. To be executed on a separate thread.
+  void spin();
```
We should refactor this function into a `spinOnce`-style callback that we can call from the `Executor` implemented in #151. That is potentially important because it lets us consolidate all of our various callbacks, e.g. multiple `RosTrajectoryExecutor`s and `JointStateSubscriber`s, into a single thread.

mkoval replied:
Nevermind! It turns out that `spin()` is already implemented that way. 😅

Maybe we should rename this to something more obvious? Maybe `step()` or `tick()`?

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+  }
+
+  // Setup the goal properties.
+  // TODO: Also set the goal properties.
```
We should _eventually_ add support for populating the `goal_tolerance`, `goal_time_tolerance`, and `path_tolerance` fields in the `control_msgs/FollowJointTrajectory` message.

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,79 @@
+{
+public:
+  RosTrajectoryExecutor(
+    ::dart::dynamics::MetaSkeletonPtr skeleton, 
```
Great point! It seems like we could replace `mSkeleton` with a list of DOFs.

gilwoolee replied:
This probably comes from my lack of knowledge in ROS. My question was, can't a `RosTrajectoryExecutor` be very flexible and take trajectories covering different set of dofs? Will `serverName` refer to something specific like arm-controller-server, etc.?

mkoval commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,290 @@
+    }
+    else
+    {
+      message << "Execution failed";
```
Believe it or not, this is actually correct! 😈 

There are two classes of errors in this function:

1. The state of the [`actionlib` state machine](http://wiki.ros.org/actionlib/DetailedDescription) is reflected by `getCommState()`. If the goal handle finishes _for any reason_ - even if communication is lost with the action server - the state machine transitions to `CommState::DONE`.
2. Once in `CommState::DONE`, `getTerminalState()` indicates why the goal terminated. This can be `TerminalState::SUCCEEDED` or one of several error conditions.
3. If the goal finished in `TerminalState::SUCCEEDED`, then `getResult()` contains the result message. In the case of `FollowJointTrajectoryResult`, the message contains an additional `error_code`.

To summarize: (1) and (2) are **`actionlib` errors** and (3) is a **`FollowJointTrajectory` error**. It's possible for the `actionlib` call to succeed, but for executing the trajectory to fail (in fact: this is the most common case of failure we observe in practice).

The `else` clause here populates `message` with "Execution failed" to differentiate `actionlib` errors from controller errors. Note that `isSuccessful` is **not** set to `false`, so the `message` will only be used if the `result->error_code != Result::SUCCESSFUL` condition fails below.

Does that make sense? Maybe add a comment or two to explain this? 😅

gilwoolee replied:
Thanks!

gilwoolee commented:
@mkoval I addressed most of your comments, which resulted in two major changes:
  - a new conversion function from `Trajectory` to `trajectory_msgs::JointTrajectory` in `Conversions.hpp`
  - a new `Exception` class that inherits from `std::runtime_error`. I am actually a bit not sure what exactly how this class will be used and what methods we need here. Your comment was:
> We should use a custom std::runtime_error subclass here. Otherwise, it will be difficult to distinguish between "expected" (e.g. trajectory was aborted) and "unexpected" failures (e.g. an invalid parameter).

Could you add to this? Should I make an enum type to classify the errors into `expected` and `unexpected`? 

Unit tests are added for the new conversion method. I still need to write unit tests for `RosTrajectoryExecutor`, but since I've already made some big changes, it'd be great if you could comment on them while I work on the tests.

@jslee02 could you take a look too? Thanks a lot!

jslee02 commented on src/control/ros/Conversions.cpp:
```diff
@@ -0,0 +1,438 @@
+    currTimeFromStart = nextTimeFromStart;
+  }
+
+  return std::move(trajectory);
```
`std::move( )` is unnecessary. Related [SO answer](http://stackoverflow.com/a/11817902).

jslee02 commented on src/control/ros/Conversions.cpp:
```diff
@@ -0,0 +1,438 @@
+    }
+
+    // For RnJoint, supports only 1D.
+    if (rnJoint && rnJoint->getDimension() > 1)
```
I think we need to explicitly check whether the dimension is not one because it can be zero.

jslee02 commented on src/control/ros/RosTrajectoryExecutor.cpp:
```diff
@@ -0,0 +1,249 @@
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr traj)
+{
+  static const ::ros::Time invalidTime;
```
Passing invalid time is intentional? I think we could simply pass the current time using `ros::Time::now()`.

mkoval replied:
Passing an invalid time **is** intentional. Passing `ros::Time::now()` computes the timestamp in this process, whereas passing `ros::Time()` causes the controller to compute the timestamp when it receives the message. 

This difference can be significant if the client and server are running on different computers.

jslee02 commented on include/aikido/control/ros/Conversions.hpp:
```diff
@@ -0,0 +1,38 @@
+/// \_param[in] trajectory Aikido trajectory to be converted.
+/// \_param[in] indexMap Mapping between trajectory's joints and ros joints.
+/// \_param[in] timestep Timestep between two consecutive waypoints.
+const trajectory_msgs::JointTrajectory convertTrajectoryToRosTrajectory(
```
I would like to suggest patterning the two convert function names. `convertJointTrajectory()` doesn't convey the original and target type in the function name unlike to the latter function name, and `convertTrajectoryToRosTrajectory()` uses only `Trajectory` to refer `JointTrajectory` unlike to the former function name.

We could put both of original and target type name to the function name like `convertJointTrajectoryToRosJointTrajectory()` but this might be too long. So my preference is including only the goal type to the function name so that we can call the correct conversion functions depending on the target type. We don't need to worry about the original type since functions can overload for the argument types, which is the original type, but not for the return type.

**TL;DR**: I suggest renaming the two function names to something like:
```cpp
convertToSplineJointTrajectory()
convertToRosJointTrajectory()

// or
toSplineJointTrajectory()
toRosJointTrajectory()
```

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutionException.hpp:
```diff
@@ -0,0 +1,33 @@
+
+  RosTrajectoryExecutionException(
+    const std::string& what,
+    actionlib::TerminalState terminalState);
```
`TerminalState` is an `enum`, so there is no point in avoiding a copy here.

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,95 @@
+    std::chrono::milliseconds connectionTimeout
+      = std::chrono::milliseconds{1000},
+    std::chrono::milliseconds connectionPollingPeriod
+      = std::chrono::milliseconds{20}
```
I am generally supportive of this, but I am not sure how to implement it. We store these `duration`s in member variables, so this would require `template`ing the entire class on `DurationA` and `DurationB`. It's possible we could do some type erasure shenaniganz to avoid this, but I don't think it's worth the effort at this point.

jslee02 replied:
We actually don't need to templatize the whole class but only this constructor. The member variable can be any specific duration type. Here is an example what I did: [member variable](https://github.com/personalrobotics/aikido/blob/master/include/aikido/util/ExecutorThread.hpp#L57), [templated constructor](https://github.com/personalrobotics/aikido/blob/master/include/aikido/util/detail/ExecutorThread-impl.hpp#L7-L11).

mkoval commented on include/aikido/control/ros/RosTrajectoryExecutor.hpp:
```diff
@@ -0,0 +1,95 @@
+
+  bool waitForServer();
+
+  void transitionCallback(GoalHandle handle);
```
Nope. This function is passed as a callback to `ActionClient::sendGoal`, which expects this signature.

mkoval commented on src/control/ros/Conversions.cpp:
```diff
@@ -0,0 +1,438 @@
+  assert(0 <= numDerivatives && numDerivatives <= 2);
+  const std::array<std::vector<double>*, 2> derivatives{
+    &waypoint.velocities, &waypoint.accelerations};
+  for (int iderivative = 1; iderivative <= numDerivatives; ++iderivative)
```
This is the convention I typically use for naming loop variables, so I suspect you're going to have a lot more comments like this on the other `ros_control` pull requests. 😅 

I generally like this convention, but I am fine with @gilwoolee renaming the variables if you insist... 😞

mkoval commented on src/control/ros/Conversions.cpp:
```diff
@@ -0,0 +1,438 @@
+void reorder(std::map<size_t, size_t> indexMap,
+  const Eigen::VectorXd& inVector, Eigen::VectorXd* outVector)
+{
+  *outVector = Eigen::VectorXd::Zero(inVector.size());
```
This is a helper function in an anonymous namespace, so I suggest just adding an `assert`.
